src/storage/observer_worker.rs

```diff
@@ -1,147 +1,218 @@
 }
 
+#[derive(Debug)]
+enum TickResult {
+    Continue,
+    Stop
+}
+
 impl<K> ObserverWorker<K>
 where
     for<'a> K: Key<'a> + 'static,
 {
     pub(crate) fn new(
         receiver: Receiver<Msg>,
-        loopback_sender: Sender<Msg>,
-        inner: Inner<K>,
-        dump_sem: Arc<Semaphore>,
-        async_oplock: Arc<Mutex<()>>,
+        inner: Arc<Inner<K>>,
     ) -> Self {
-        let deferred_min_duration = inner.config.deferred_min_time().clone();
-        let deferred_max_duration = inner.config.deferred_max_time().clone();
         Self {
             inner,
             receiver,
-            loopback_sender,
-            dump_sem,
-            async_oplock,
-            deferred: Arc::default(),
-            deferred_min_duration,
-            deferred_max_duration,
+            next_deadline: None,
+            deferred_index_dump_info: None,
+            index_dump_task: None
         }
     }
 
     pub(crate) async fn run(mut self) {
         loop {
-            if let Err(e) = self.tick().await {
-                warn!("active blob will no longer be updated, shutdown the system");
-                warn!("{}", e);
-                break;
+            if self.index_dump_task.as_ref().map_or(false, |task| task.is_finished()) {
+                // Complete task if it is already finished
+                complete_task(&mut self.index_dump_task, "index_dump_task").await;
+            }
+
+            let tick_result = 
+                match self.next_deadline {
+                    None => self.tick().await,
+                    Some(deadline) => self.tick_with_deadline(deadline).await
+                };
+
+            match tick_result {
+                Ok(TickResult::Continue) => {},
+                Ok(TickResult::Stop) => {
+                    debug!("ObserverWorker stopping. No future update is possible");
+                    break;
+                },
+                Err(err) => {
+                    error!("ObserverWorker unexpected error: {:?}", err);
+                    panic!("ObserverWorker unexpected error: {:?}", err);
+                }
             }
         }
-        info!("observer stopped");
+
+        // Wait for background task completion
+        complete_task(&mut self.index_dump_task, "index_dump_task").await;
+
+        debug!("observer stopped");
     }
 
-    async fn tick(&mut self) -> Result<()> {
+    async fn tick(&mut self) -> Result<TickResult> {
         match self.receiver.recv().await {
-            Some(msg) => self.process_msg(msg).await,
+            Some(msg) => {
+                self.process_msg(msg).await?;
+                Ok(TickResult::Continue)
+            },
+            None => Ok(TickResult::Stop)
+        }
+    }
+
+    async fn tick_with_deadline(&mut self, deadline: Instant) -> Result<TickResult> {
+        // Extend deadline a little bit to guaranty, that process_defered will detect exceeding
+        let deadline = deadline + Duration::from_millis(10);

```

Comment:
Shouldn't this duration depend on config's defered min/max time?

Reply:
Yes, the `deadline` is updated in `deffer_blob_indexes_dump` and in `process_deferred_blob_index_dump` according to  deferred min/max time from config.

The idea of this change is to update the `deadline`, every time `DeferredDumpBlobIndexes` message income and on every loop, when deadline is not `None` runs `recv` with additional timeout. If timeout triggered, run processing of deferred index dump and reset `deadline` value to `None.

Reply:
But `Duration::from_millis(10)` does not depend on config values, and therefore can be too large/too small in some situations, if I understood correctly

Reply:
The idea is to set the duration a little longer than the original one, so that the check in `process_deferred_blob_index_dump` is guaranteed to be successful.

> But Duration::from_millis(10) does not depend on config values, and therefore can be too large/too small in some situations, if I understood correctly

The original concept of deferred execution is that the processing will happen at some time later. If it happens an additional 10ms later, this is not critical. The initial values in the config are measured in minutes.

Actually, I think, it would be good to reduce `10ms` to `1ms`